### PR TITLE
TEF-663 Raise NullFactError when a fact resolves to nil

### PR DESCRIPTION
**Pull request by dfl**
## Ticket
https://codeforamerica.atlassian.net/browse/TEF-663

## Summary

Enforces a core fact-graph invariant: **no fact may resolve to `nil`.** When a value can't be calculated, a fact must signal an unmet dependency (an error hash / `data_errors`) so the facts downstream can tell a *missing* value apart from a *satisfied* one. A fact that returns `nil` breaks that guarantee silently — dependents treat the gap as answered.

`Fact#call` now raises `FactGraph::NullFactError` — naming the offending fact (and entity, for per-entity facts) — whenever a resolver or constant resolves to `nil`.

## Why

This is a developer-error invariant, like a `NoMethodError`: a `nil` fact is a bug in the graph definition, so it should fail loudly rather than propagate. Catching it at evaluation time covers **every** input, everywhere the graph runs (dev, test, prod, jobs) — not just paths a spec happens to exercise. It closes the two ways `nil` leaks in:

- an `allow_unmet_dependencies` fact whose branch returns `nil` instead of `data_errors`
- an input fact whose validator permits `nil` (the blank case should resolve to an error, not pass `nil` through)

`false` remains a valid value and does not raise.

## Changes

- `lib/fact_graph.rb` — add `FactGraph::NullFactError`
- `lib/fact_graph/fact.rb` — `ensure_non_nil!` guard on both the resolver and constant/pass-through paths
- `spec/fact_graph/fact_spec.rb` — cover the raise (proc, `allow_unmet_dependencies`, input pass-through, per-entity message) and that `false` does **not** raise
- `README.md` — document the invariant under "Handling unmet dependencies"

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: lib/fact_graph/fact.rb [hunk 1]
```diff
@@ -147,6 +147,7 @@ def call(input, results)
     results[module_name] ||= {}
 
     if !resolver.respond_to?(:call)
```
Comment (dfl) [inline lib/fact_graph/fact.rb:149]: thoughts on flipping this to `unless` ?

File: spec/fact_graph/fact_spec.rb [hunk 2]
```diff
@@ -175,6 +175,79 @@
...
+    it "raises when an input fact passes through a nil value" do
+      Class.new(FactGraph::Graph) do
+        in_module :leaks do
+          fact :optional do
+            input [:wrapper, :optional]
+          end
+        end
+      end
+      expect { evaluate_single_fact(:leaks, :optional, {wrapper: {optional: nil}}) }
+        .to raise_error(FactGraph::NullFactError, /:leaks\/:optional resolved to nil/)
```
Comment (mrotondo) [inline spec/fact_graph/fact_spec.rb]: interesting. this implies that we would get a bunch of these exceptions in gyraffe _if_ we left an input fact with no validations - i.e. the validations failing on `nil` will prevent us from from having this exception raised by facts like
```
fact :middle_initial do
  input [:primary, :middle_initial], value: :string, format?: /^[a-zA-Z]$/
end
```
when elsewhere in the input adapter we just say
```
def dependent_input(dependent)
  {
...
    middle_initial: dependent.middle_initial,
...
  }
end
```

no action or changes necessary, just something for us to look out for since "what is the value of our facts' input validations?" has been in the conversation lately

File: spec/fact_graph/fact_spec.rb [hunk 3]
```diff
@@ -175,6 +175,79 @@
...
+      Class.new(FactGraph::Graph) do
+        in_module :leaks do
+          fact :leaky do
+            proc { nil }
+          end
+        end
+      end
+      expect { evaluate_single_fact(:leaks, :leaky) }
+        .to raise_error(FactGraph::NullFactError, /:leaks\/:leaky resolved to nil/)
+    end
```
Comment (mrotondo) [inline spec/fact_graph/fact_spec.rb]: all these tests exercise facts with callable evaluators. can you add another test case on a fact with a non-callable resolver that returns nil, i.e. `fact :nil_constant { nil }`?

**Review by mrotondo — APPROVED** [review]
looks good! a non-blocking request for one additional test case for the un-covered early return branch for non-callable resolvers, and a note about how this might behave when we get it into gyraffe.

File: lib/fact_graph/fact.rb (hunk 1 above)
Comment (mrotondo) [reply lib/fact_graph/fact.rb]: fine with that. i try not to have strong feelings on `unless` even though i tend to have to think about it a few seconds longer
